lightning-block-sync/src/lib.rs

```diff
@@ -0,0 +1,90 @@
+//! A lightweight client for keeping in sync with chain activity.
+//!
+//! Defines a [`BlockSource`] trait, which is an asynchronous interface for retrieving block headers
+//! and data.
+//!
+//! Enabling feature `rest-client` or `rpc-client` allows configuring the client to fetch blocks
+//! using Bitcoin Core's REST or RPC interface, respectively.
+//!
+//! Both features support either blocking I/O using `std::net::TcpStream` or, with feature `tokio`,
+//! non-blocking I/O using `tokio::net::TcpStream` from inside a Tokio runtime.
```

Comment:
I may have added to the confusion during our discussion. For the non-tokio case, there is never an `await` on stream I/O since the stream used (`std::net::TcpStream`) has a synchronous interface. Therefore, stream I/O will block the thread.

For the tokio case, `tokio::net::TcpStream` has an asynchronous interface (i.e., it requires calling `await` on the returned future). Whenever steam I/O is hit, the runtime is given the chance to schedule other work if blocking would have occurred. Therefore, stream I/O will not block the thread.

So while the overall `BlockSource` interface is asynchronous (i.e., written in terms of `async` functions), `HttpClient` is only non-blocking when using the `tokio` feature. But either way a runtime is needed to drive the execution.

Let me know if that cleared anything up and if you have any suggestions for making the docs more intuitive.